src/main/scala/tech/sourced/gemini/Gemini.scala

```diff
@@ -58,30 +61,52 @@ object Gemini {
       //TODO: implement based on Apolo
       //findSimilarProjects(path, conn)
     } else {
-      findDuplicateFiles(path, conn)
+      findDuplicateBlobHash(path, conn)
       //TODO: implement based on Apolo
       //findSimilarFiles(path, conn)
     }
   }
 
   /**
-    * Finds duplicated files among hashed repositories that already exists in certain repository
+    * Finds duplicate files among hashed repositories
+    *
+    * @param conn   Database connection
+    * @return
+    */
+  def report(conn: Session): Iterable[DuplicateBlobHash] = {
+    findDuplicateBlobHash(conn)
+  }
+
+  /**
+    * Finds duplicate files among hashed repositories, and returns all duplicated files
     *
-    * @param repository repository url, example: github.com/src-d/go-git.git"
+    * @param conn   Database connection
     * @return
     */
-  def report(repository: String): Iterable[RepoFile] = {
-    throw new UnsupportedOperationException("Finding all duplicate files in many repositories is no implemented yet.")
+  def detailedReport(conn: Session): Iterable[Iterable[RepoFile]] = {
+    findDuplicateBlobHash(conn)
+        .map {d => findDuplicateBlobHash(d.sha, conn)}
+  }
+
+  def findDuplicateBlobHash(conn: Session): Iterable[DuplicateBlobHash] = {
+    val cql = "SELECT blob_hash, repo, file_path, COUNT(*) as count FROM hashes.blob_hash_files GROUP BY blob_hash"
+    conn
+      .execute(new SimpleStatement(cql))
```

Comment:
usually it's a good idea to extract constants with the queries as a top level fields in object/class.